lib/ansible/modules/windows/win_psrepository.ps1

```diff
@@ -1,68 +1,229 @@
 $check_mode = Get-AnsibleParam -obj $params -name "_ansible_check_mode" -type "bool" -default $false
 
 $name = Get-AnsibleParam -obj $params -name "name" -type "str" -failifempty $true
-$source = Get-AnsibleParam -obj $params -name "source" -type "str"
+$source_location = Get-AnsibleParam -obj $params -name "source_location" -aliases "source","module_source","module_source_location" -type "str"
+$script_source_location = Get-AnsibleParam -obj $params -name "script_source_location" -aliases "script_source" -type "str"
+$publish_location = Get-AnsibleParam -obj $params -name "publish_location" -aliases "module_publish_location" -type "str"
+$script_publish_location = Get-AnsibleParam -obj $params -name "script_publish_location" -type "str"
 $state = Get-AnsibleParam -obj $params -name "state" -type "str" -default "present" -validateset "present", "absent"
-$installationpolicy = Get-AnsibleParam -obj $params -name "installation_policy" -type "str" -validateset "trusted", "untrusted"
+$installation_policy = Get-AnsibleParam -obj $params -name "installation_policy" -type "str" -validateset "trusted", "untrusted"
+$resolve_locations = Get-AnsibleParam -obj $params -name "resolve_locations" -type "bool" -default $true
+$force = Get-AnsibleParam -obj $params -name "force" -type "bool" -default $false
 
 $result = @{"changed" = $false}
 
+Function Resolve-PSGetLocation {
+    <#
+        .SYNOPSIS
+        This is a sort of proxy function for the Resolve-Location function that PowerShellGet uses internally.
+
+        .DESCRIPTION
+        It's not exported from the module, and it needs a caller context so wrapping it in a function both convenient
+        and functional depending on how the outer module is executed (where $PSCmdlet may not be set).
+
+        Also wraps the needed to call to other internal function, Get-LocationString, which does some normalizing
+        or [uri] to [string] conversion.
+
+        .NOTES
+        - This must be an advanced function so that $PSCmdlet is populated.
+        - Some context around all the things this tries to do: https://github.com/PowerShell/PowerShellGet/issues/317
+    #>
+    [CmdletBinding()]
+    [OutputType([string])]
+    Param
+    (
+        [Parameter(Mandatory=$true)]
+        [uri]
+        $Location,
+
+        [Parameter(Mandatory=$true)]
+        [string]
+        $LocationParameterName,
+
+        [Parameter()]
+        [pscredential]
+        $Credential,
+
+        [Parameter()]
+        $Proxy,
+
+        [Parameter()]
+        [pscredential]
+        $ProxyCredential
+    )
+
+    $module = Get-Module -Name PowerShellGet
+
+    $module.Invoke({
+        param(
+            [System.Collections.IDictionary]
+            $Params,
+
+            [System.Management.Automation.PSCmdlet]
+            $CallerPSCmdlet
+        )
+        $Params.Location = Get-LocationString -LocationUri $Params.Location
+        Resolve-Location @Params -CallerPSCmdlet $CallerPSCmdlet
```

Comment:
[Resolve-Location](https://github.com/PowerShell/PowerShellGet/blob/development/src/PowerShellGet/private/functions/Resolve-Location.ps1) looks like it might be a private function I would be very reluctant to rely on that in our modules. What is it meant to be doing? Trying to get the final URL on a redirected one?

Reply:
Yes it's private, I wrote info about it in the XML help.

I originally built a function that followed HTTP/S redirects. It was working for that purpose but I found that it failed to produce the same URL as PowerShellGet in some cases and that's when I found out that this internal function tries to do other stuff, like probe for NuGet API responses and add on versioned API extensions to the URL. It also does some minor changes to paths in filesystem repos. All in all, it seemed like an awful lot to be duplicating the logic of it all. 

I think relying on the internal function gives us the most consistent outcome, especially if that function changes in future versions as the result of `Resolve-Location` will always be what PowerShellGet itself is writing out. 

The next best alternative is probably to copy and paste it out of the source code and duplicate it here, which protects us from the name changing, but leaves us more vulnerable to the implementation changing.

Open to suggestions on alternatives

Reply:
There's nothing stopping PSGet from changing/moving/removing this cmdlet in future updates effectively breaking this module. If you wanted to fix this particular problem I would just try and replicate the same behaviour as PSGet and doing a GET (HEAD might even be better) and returning the location if it was redirected https://github.com/PowerShell/PowerShellGet/blob/development/src/PowerShellGet/private/functions/Ping-Endpoint.ps1#L107.

You can even utilise the [Ansible.ModuleUtils.WebRequest](https://github.com/ansible/ansible/blob/devel/lib/ansible/module_utils/powershell/Ansible.ModuleUtils.WebRequest.psm1). You don't need to expose the whole share parameters aspect but you can utilise `Get-AnsibleWebRequest` and `Invoke-AnsibleWebRequest` with `-IgnoreBaseResponse` to do a similar thing as PSGet. Feel free to reach out to me if you need some more help on trying to get that set up. We do something similar in win_uri so that's a good place to start. If you do go down that route I wouldn't expose those parameters as module options, just pass in whatever you need normally.

Reply:
I don't know if it's still viewable now, I may have squashed the commits (I stashed the redirect following code I had somewhere else if you want to see it), but I really did go down that road before, even used `HEAD` requests, until it broke in the tests because myget.org doesn't support `HEAD` requests, so I added fallback to `GET`, and continued on until running into the next issue which is that `Resolve-Location` doesn't just follow 3xx redirects. So there's more to add if we were going to try to emulate that behavior.

Highly recommended reading as this shed a lot of light on what a minefield this is: https://github.com/PowerShell/PowerShellGet/issues/317

If it's more desirable to re-implement their logic than copying their code directly (it's MIT licensed) or relying on the private method, I think it might be a better idea to leave that to a new PR and keep that as a separate issue. 

My main purpose was to add support for the other location parameters; I just happened to find this bug during the course of that, but it's turned out to be far more time consuming than expected to address and is probably best solved on its own.

What do you think of a compromise for the time being where if the private method is not found or fails, we issue a warning and continue on? The result in that scenario would be the same behavior that's in place now (but with notice) in event that something internal changes, but in the meantime it would be more reliable (this would still be leaving #67504 open so that it can be addressed permanently in a new PR).

Reply:
Totally understand that this is not an ideal situation all round and PowerShellGet has a lot of quirks that make these modules difficult to deal with. In saying that by trying to re-utilise the `Resolve-Location` cmdlet which isn't even exposed in the modules manifest you are relying on an internal implementation that is not guaranteed to be there in future releases. Considering there is a v3 release in the works that is meant to solve a lot of reliability issues of PSGet I would wager that this function has gone the way of the dodo effectively breaking our module if we took the changes as is.

I wouldn't want to copy their code, I've been burnt too many times with PSGet to consider having their code in these modules and even just trying to decipher their current implementation is a bit mind numbing. Ultimately I think if you wanted to try and solve this particular issue (and if I understand the problem correctly) you can solve it with something like

```
Function Resolve-Location {
    [CmdletBinding()]
    param (
        [Parameter(Mandatory=$true)]
        [Uri]
        $Location,

        [Parameter(Mandatory=$true)]
        [Object]
        $Module
    )

    if ($Location.Scheme -notin @('http', 'https')) {
        $Location
        return
    }

    $request = Get-AnsibleWebRequest -Uri $Location -Method GET -FollowRedirects All
    try {
        Invoke-WithWebRequest -Request $request -IgnoreBadResponse -Module $Module -Script {
            Param ([System.Net.WebResponse]$Response, [System.IO.Stream]$Stream)
    
            $Response.ResponseUri.AbsoluteUri
        }
    } catch [System.Net.WebException] {
        # Probably unable to resolve the URL, just ignore this and rely on PSGet to fail or not.
        $Location
    }
}

$source_location = Resolve-Location -Location $source_location -Module $module
$publish_location = Resolve-Location -Location $publish_location -Module $module
...
```

By using this util you take advantage of all the complexities when it comes to dealing web requests in PowerShell such as

* Remembering to enable TLS 1.2
* Easily set the timeout or redirection settings if desired
* If we wanted to add support for proxies or web auth, this is all tested and done for you in the util

A quick test shows that this work for me and seems to be similar to what PSGet does, just with less lines.

Reply:
That solves for 3xx redirects but not the other modifications that `Resolve-Location` does (v2/v3 API checks, filesystem path changes). 

I definitely agree that relying on a private function is not ideal and may be prone to failure, so I understand not wanting to take a dependency on it.

Given the complexity I think I'd like to separate this issue into a new PR that directly addresses it (maybe after the split), and keep this PR scoped down to adding the new locations and `force` support.

In the new PR I can work toward using the Ansible web request module util like you have above.

Reply:
That sounds like a great plan. Let’s get in support for the new parameters and not have this hold up the PR.